Problem: TaskRunner panics on commit message with non-ASCII byte at position 100 (UTF-8 truncation)

### Issue

## Summary

A non-ASCII character at byte position 100 of the HEAD commit message kills the task worker. Every queued task on that branch panics on startup until the offending commit is no longer HEAD.

## Reproducer

```bash
git commit --allow-empty -m "$(printf '%.0sa' {1..99})ы"
git push
```

99 ASCII bytes + 2-byte `ы`. Trigger any task on this branch in Semaphore.

## Actual

```
task: non-zero exit (2) - time="2026-05-04T15:47:06Z" level=panic msg="pq: invalid byte sequence for encoding \"UTF8\": 0xd1 (22021)" error="Failed to update TaskRunner status"

panic: (*logrus.Entry) 0x400060a690
    panic: (*logrus.Entry) 0x400060a9a0
```

## Expected

Task runs. Commit message stored as valid UTF-8 truncated to fit `VARCHAR(100)`.

## Workaround

Push another commit with an ASCII-only subject.

## Environment

- Semaphore community (`develop` and current release)
- PostgreSQL with `client_encoding = UTF8` (default)
- Both `cmd` and `go-git` repo clients reproduce

### Impact

Docker, Web-Backend (APIs), Service (scheduled tasks, alerts)

### Installation method

Docker

### Database

Postgres

### Browser

_No response_

### Semaphore Version

v2.17.39

### Ansible Version

```bash

```

### Logs & errors

```
task: non-zero exit (2) - time="2026-05-04T15:47:06Z" level=panic msg="pq: invalid byte sequence for encoding \"UTF8\": 0xd1 (22021)" error="Failed to update TaskRunner status"

panic: (*logrus.Entry) 0x400060a690
    panic: (*logrus.Entry) 0x400060a9a0

goroutine 522 [running]:
github.com/sirupsen/logrus.(*Entry).log(0x400060a930, 0x0, {0x40000f8e80, 0x3b})
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:263 +0x490
github.com/sirupsen/logrus.(*Entry).Log(0x400060a930, 0x0, {0x40006d2188?, 0x5?, 0x10?})
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:307 +0x60
github.com/sirupsen/logrus.(*Entry).Panic(...)
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:345
github.com/semaphoreui/semaphore/util.LogPanicF({0x2897c20, 0x4000001200}, 0xe73fb7?)
    /go/src/semaphore/util/errorLogging.go:46 +0x80
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).panicOnError(0x21?, {0x2897c20, 0x4000001200}, {0xe9cbed, 0x22})
    /go/src/semaphore/services/tasks/TaskRunner_logging.go:150 +0x100
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).saveStatus(0x4000546008)
    /go/src/semaphore/services/tasks/TaskRunner.go:103 +0x578
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).run.func1()
    /go/src/semaphore/services/tasks/TaskRunner.go:185 +0x19c
panic({0xe43520?, 0x400060a690?})
    /usr/local/go/src/runtime/panic.go:792 +0x124
github.com/sirupsen/logrus.(*Entry).log(0x400060a620, 0x0, {0x40000f8c80, 0x3b})
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:263 +0x490
github.com/sirupsen/logrus.(*Entry).Log(0x400060a620, 0x0, {0x40006d2ac8?, 0x5?, 0x10?})
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:307 +0x60
github.com/sirupsen/logrus.(*Entry).Panic(...)
    /go/src/semaphore/vendor/github.com/sirupsen/logrus/entry.go:345
github.com/semaphoreui/semaphore/util.LogPanicF({0x2897c20, 0x40000005a0}, 0xe73fb7?)
    /go/src/semaphore/util/errorLogging.go:46 +0x80
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).panicOnError(0x21?, {0x2897c20, 0x40000005a0}, {0xe93d5a, 0x1c})
    /go/src/semaphore/services/tasks/TaskRunner_logging.go:150 +0x100
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).SetCommit(0x4000546008, {0x400083c000, 0x28}, {0x4000520000, 0x64})
    /go/src/semaphore/services/tasks/TaskRunner_logging.go:78 +0x13c
github.com/semaphoreui/semaphore/services/tasks.(*LocalJob).SetCommit(...)
    /go/src/semaphore/services/tasks/LocalJob.go:71
github.com/semaphoreui/semaphore/services/tasks.(*LocalJob).checkoutRepository(0x4000546708)
    /go/src/semaphore/services/tasks/LocalJob.go:987 +0x344
github.com/semaphoreui/semaphore/services/tasks.(*LocalJob).prepareRun(0x4000546708, {{0x0, 0x0, 0x0}, {0xc889e0, 0x400059c410}, {0xc889a0, 0x40005968a0}, {0x7f77a96e98, 0x40002f9040}})
    /go/src/semaphore/services/tasks/LocalJob.go:832 +0x374
github.com/semaphoreui/semaphore/services/tasks.(*LocalJob).Run(0x4000546708, {0x4000648ce0, 0x5}, 0x0, {0x0, 0x0})
    /go/src/semaphore/services/tasks/LocalJob.go:712 +0x5ac
github.com/semaphoreui/semaphore/services/tasks.(*TaskRunner).run(0x4000546008)
    /go/src/semaphore/services/tasks/TaskRunner.go:218 +0x38c
github.com/semaphoreui/semaphore/services/tasks.runTask.func1()
    /go/src/semaphore/services/tasks/TaskPool.go:362 +0x34
created by github.com/semaphoreui/semaphore/services/tasks.runTask in goroutine 49
    /go/src/semaphore/services/tasks/TaskPool.go:360 +0x3a8
```

### Manual installation - system information

_No response_

### Configuration

_No response_

### Additional information

_No response_

--- a/db_lib/CmdGitClient.go
+++ b/db_lib/CmdGitClient.go
@@ -155,9 +155,7 @@ func (c CmdGitClient) GetLastCommitMessage(r GitRepository) (msg string, err err
 		return
 	}
 
-	if len(msg) > 100 {
-		msg = msg[0:100]
-	}
+	msg = truncateCommitMessage(msg)
 
 	return
 }

--- a/db_lib/GoGitClient.go
+++ b/db_lib/GoGitClient.go
@@ -230,10 +230,7 @@ func (c GoGitClient) GetLastCommitMessage(r GitRepository) (msg string, err erro
 		return
 	}
 
-	msg = headCommit.Message
-	if len(msg) > 100 {
-		msg = msg[0:100]
-	}
+	msg = truncateCommitMessage(headCommit.Message)
 
 	r.Logger.Log("Message: " + msg)
 

--- a/db_lib/commit_message.go
+++ b/db_lib/commit_message.go
@@ -1,0 +1,13 @@
+package db_lib
+
+// Matches `commit_message VARCHAR(100)` in db/sql/migrations.
+const commitMessageMaxRunes = 100
+
+// Slice by runes — byte slicing can split a multi-byte UTF-8 sequence.
+func truncateCommitMessage(msg string) string {
+	runes := []rune(msg)
+	if len(runes) > commitMessageMaxRunes {
+		return string(runes[:commitMessageMaxRunes])
+	}
+	return msg
+}

--- a/db_lib/commit_message_test.go
+++ b/db_lib/commit_message_test.go
@@ -1,0 +1,52 @@
+package db_lib
+
+import (
+	"strings"
+	"testing"
+	"unicode/utf8"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTruncateCommitMessage_KeepsFullString(t *testing.T) {
+	tests := []struct {
+		name  string
+		input string
+	}{
+		{"empty", ""},
+		{"short ASCII", "fix: short"},
+		{"100 ASCII", strings.Repeat("a", 100)},
+		{"100 Cyrillic", strings.Repeat("ы", 100)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.Equal(t, tt.input, truncateCommitMessage(tt.input))
+		})
+	}
+}
+
+func TestTruncateCommitMessage_Truncates(t *testing.T) {
+	tests := []struct {
+		name  string
+		input string
+		want  string
+	}{
+		{"101 ASCII", strings.Repeat("a", 101), strings.Repeat("a", 100)},
+		{"150 Cyrillic", strings.Repeat("ы", 150), strings.Repeat("ы", 100)},
+		{"150 emoji", strings.Repeat("🙂", 150), strings.Repeat("🙂", 100)},
+		{"99 ASCII + 3 Cyrillic", strings.Repeat("a", 99) + "ыыы", strings.Repeat("a", 99) + "ы"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.Equal(t, tt.want, truncateCommitMessage(tt.input))
+		})
+	}
+}
+
+func TestLegacyByteSliceProducesInvalidUTF8(t *testing.T) {
+	msg := "Hello мир, γειά κόσμε, مرحبا بالعالم, שלום עולם, 你好世界, こんにちは 안녕하세요 🌍🎉"
+
+	require.Greater(t, len(msg), 100)
+	assert.False(t, utf8.ValidString(msg[:100]))
+}

